thinc/layers/concatenate.py

```diff
@@ -72,12 +74,12 @@ def backprop(d_output: Array2d) -> InT:
             start += width
         return dX
 
-    return output, backprop
+    return cast(OutT, output), backprop
 
 
 def _ragged_forward(
-    model: Model[InT, Ragged], X, Ys, callbacks, is_train: bool
-) -> Tuple[Ragged, Callable]:
+    model: Model[InT, OutT], X, Ys, callbacks, is_train: bool
+) -> Tuple[OutT, Callable]:
```

Comment:
I'm not sure I understand why the types have been changed to these private functions. Wasn't it correct?

Reply:
You can do it either way, but these are private functions so it's really an implementation detail. I think I ended up doing it like this because it meant the `cast`s were closer to where the data was actually being generated and the main `forward` method was clearer, but I don't think it really makes much difference.

Reply:
I agree that the main function would be the better place for the functionality, but now trying to actually move it there I see what the problem is: you need two casts for each subtype, one going into the helper function and one coming back out of it, and because what each helper function returns is a tuple where only the first element needs to be cast, the two elements of the tuple have to be assigned to separate variables and then the first element cast before it is returned. So doing it this way makes the code much more complicated. This is presumably why the old version has `#type: ignore` statements to which I think this solution is to be preferred.

Reply:
I can see that the main method would be more logically appropriate, but having just tried it's essentially impossible without severely bloating the main method. For each helper method call the function parameter, the model itself and the return type all have to be cast separately. Because the helper methods return tuples and only the first member of each tuple has to be cast, the return values from each helper method have to be assigned to variables before the first one can be cast. And unlike in `dropout` where the return type is bound to a parameter of the main method, the variables used for each return type would have to have different names. I think these problems are why the original code uses `#type: ignore` and find that this solution is the worse of two evils.

Reply:
This is a case where functionality should be more important than aesthetics though. Functionally, the private helper functions simply shouldn't know about the broader typing. They are created specifically to deal with only one type, so they should be typed as such.

Reply:
I can see that and I'm changing all the layers as you suggest. The downside though is less aesthetics than quite significant changes to the actual code (rather than just types) which I had been trying to avoid in case something breaks. However, now we're not at the end of the release cycle this point shouldn't be quite so critical.